### PR TITLE
Warn when implicit is enclosing owner

### PR DESCRIPTION
**Pull request by som-snytt**
Maybe put it behind a flag?

Fixes scala/bug#10760

### CODE REVIEW

File: test/files/neg/implicitly-self.check [hunk 1]
```diff
@@ -0,0 +1,12 @@
+implicitly-self.scala:3: warning: Implicit resolves to enclosing symbol method c
+  implicit def c: Char = implicitly[Char]
+                                   ^
+implicitly-self.scala:4: warning: Implicit resolves to enclosing symbol value s
+  implicit val s: String = implicitly[String]
+                                     ^
+implicitly-self.scala:6: warning: Implicit resolves to enclosing symbol value t
```
Comment (hrhino) [inline test/files/neg/implicitly-self.check]: I would bikeshed for `resolves to enclosing value t`, &c.

**Review by hrhino — COMMENTED** [review]
A related case, only dubiously plausible:

```scala
scala> trait TC[T] { def ix: Int }
defined trait TC

scala> implicit object tcString extends TC[String] { def ix = implicitly[TC[String]].ix + 1 }
defined object tcString

scala> tcString.ix
java.lang.StackOverflowError
  at tcString$.ix(<console>:18)
  at tcString$.ix(<console>:18)
  at tcString$.ix(<console>:18)
```

File: test/files/neg/implicitly-self.check (hunk 1 above)
Comment (som-snytt) [reply test/files/neg/implicitly-self.check]: Yep.

**Comment by som-snytt** [comment]
Mr Abot's output says something about `MAX_SURVIVABILITY`.

Is "Mr Abot" an alias for SkyNet?

**Review by lrytz — APPROVED** [review]
Thanks!

**Comment by lrytz** [comment]
@hrhino please also feel free to use GitHub's review feature and accept a PR!

**Comment by jvican** [comment]
This is a nice UX improvement, thanks @som-snytt :smile:

**Comment by som-snytt** [comment]
The next improvement is to turn it on only in a context where it could be triggered. Most implicit searches are not done from an implicit def.

**Comment by som-snytt** [comment]
I needed this today! Because of the issue about simplifying implicits, I was experimenting and accidentally wrote the classic `implicit val f: String => Int = _.toInt`. Also, the warning is broken for conversions.

```
$ skala -Ywarn-self-implicit -language:_
Welcome to Scala 2.12.5 (OpenJDK 64-Bit Server VM 1.8.0_151)

scala> implicit val f: String => Int = s => {implicitly[String => AnyVal { def toInt: Int } ]}.apply(s).toInt
<console>:18: warning: Implicit resolves to enclosing value f
       implicit val f: String => Int = s => {implicitly[String => AnyVal { def toInt: Int } ]}.apply(s).toInt
                                                       ^
f: String => Int = $$Lambda$2074/523699314@50d91a0f

scala> implicit val f: String => Int = _.toInt
f: String => Int = $$Lambda$2033/380881157@49fdbe2b

scala> implicit val f: String => Int = _ => ""
f: String => Int = $$Lambda$2021/1835105031@340c57e0

scala> f("42")
java.lang.StackOverflowError

```

**Pull request merged**